extensions/Worker.Extensions.CosmosDB/src/Config/CosmosDBBindingOptions.cs

```diff
@@ -32,31 +40,33 @@ internal virtual CosmosClient GetClient(string preferredLocations = "")
                 throw new ArgumentNullException(nameof(ConnectionName));
             }
 
-            string cacheKey = BuildCacheKey(ConnectionName!, preferredLocations);
-
-            CosmosClientOptions cosmosClientOptions = new ()
+            if (CosmosExtensionOptions is null)
             {
-                ConnectionMode = ConnectionMode.Gateway
-            };
+                CosmosExtensionOptions = new CosmosDBExtensionOptions();
+            }
+
+            string cacheKey = BuildCacheKey(ConnectionName!, preferredLocations);
 
-            if (!string.IsNullOrEmpty(preferredLocations))
+            // Do not override if preferred locations is configured via CosmosClientOptions
+            if (!string.IsNullOrEmpty(preferredLocations) && CosmosExtensionOptions.ClientOptions.ApplicationPreferredRegions is null)
             {
-                cosmosClientOptions.ApplicationPreferredRegions = Utilities.ParsePreferredLocations(preferredLocations);
+                CosmosExtensionOptions.ClientOptions.ApplicationPreferredRegions = Utilities.ParsePreferredLocations(preferredLocations);
```

Comment:
Because preferred locations are supplied per function, we can't mutate the global options here. That will impact other bindings with potentially different preferred locations.

We will need to apply preferred locations without impacting other functions.

- Can we make a copy of the options object?
- Can we use named options? Using the function name as the key? Then we will have a copy of options per function. This is a bit bigger of a scope creep for this PR though.
- Some other way to avoid this mutation?

Reply:
Great callout here, I went with the copy approach.